bindings/matrix-sdk-ffi/src/api.udl

```diff
@@ -112,57 +106,179 @@ interface Room {
+    InsertAtData? insert_at();
+    [Self=ByArc]
+    UpdateAtData? update_at();
+    u32? remove_at();
+    MoveData? move();
+    [Self=ByArc]
+    TimelineItem? push();
 };
 
-interface BaseMessage {
-    string id();
-    string body();
+enum TimelineChange {
+    "Replace",
+    "InsertAt",
+    "UpdateAt",
+    "RemoveAt",
+    "Move",
+    "Push",
+    "Pop",
+    "Clear",
+};
+
+dictionary InsertAtData {
+    u32 index;
+    TimelineItem item;
+};
+
+dictionary UpdateAtData {
+    u32 index;
+    TimelineItem item;
+};
+
+dictionary MoveData {
+    u32 old_index;
+    u32 new_index;
+};
+
+interface TimelineItem {
+    [Self=ByArc]
+    EventTimelineItem? as_event();
+    [Self=ByArc]
+    VirtualTimelineItem? as_virtual();
+};
+
+interface EventTimelineItem {
+    TimelineKey key();
     string sender();
+    TimelineItemContent content();
+    sequence<Reaction> reactions();
     u64 origin_server_ts();
-    string? transaction_id();
 };
 
-interface TextMessage {
-    BaseMessage base_message();
-    string? html_body();
+[Enum]
+interface TimelineKey {
+  LocalOnly(string txn_id);
+  Synced(string event_id);
 };
 
-interface ImageMessage {
-    BaseMessage base_message();
-    MediaSource source();
-    u64? width();
-    u64? height();
-    string? blurhash();
+interface TimelineItemContent {
+    [Self=ByArc]
+    Message? as_message();
+    boolean is_redacted_message();
 };
 
-interface NoticeMessage {
-    BaseMessage base_message();
-    string? html_body();
+interface Message {
+    MessageType msgtype();
+    // This event ID string will be replaced by something more useful later.
+    string? in_reply_to();
+    boolean is_edited();
 };
 
-interface EmoteMessage {
-    BaseMessage base_message();
-    string? html_body();
+interface MessageType {
+    string body();
+
+    EmoteMessageContent? as_emote();
+    ImageMessageContent? as_image();
+    NoticeMessageContent? as_notice();
+    TextMessageContent? as_text();
+};
+
+dictionary EmoteMessageContent {
+    string body;
+    FormattedBody? formatted;
+};
+
+dictionary ImageMessageContent {
+    string body;
+    MediaSource source;
+    ImageInfo? info;
+};
+
+dictionary ImageInfo {
+    u64? height;
+    u64? width;
+    string? mimetype;
+    u64? size;
+    ThumbnailInfo? thumbnail_info;
+    MediaSource? thumbnail_source;
+    string? blurhash;
+};
+
+dictionary ThumbnailInfo {
+    u64? height;
+    u64? width;
+    string? mimetype;
+    u64? size;
+};
+
+dictionary NoticeMessageContent {
+    string body;
+    FormattedBody? formatted;
+};
+
+dictionary TextMessageContent {
+    string body;
+    FormattedBody? formatted;
 };
 
+dictionary FormattedBody {
+    MessageFormat format;
+    string body;
+};
+
+enum MessageFormat {
+    "Html",
+    "Unknown",
+};
+
+dictionary Reaction {
+    string key;
+    sequence<ReactionDetails> details;
+};
+
+dictionary ReactionDetails {
+    TimelineKey id;
+    string sender;
+};
+
+interface VirtualTimelineItem {};
```

Comment:
I'm guessing this is intentionally not exposed yet, but just pointing out that it doesn't provide anything useable yet.

Reply:
Yeah, this is just a placeholder. It doesn't make much of a difference whether I expose it until the timeline actually emits these items :)